TID (OID) Configurator (#844)
opengisch/QgisModelBaker: docs/background_info/oid_tid_generator.md

```
[...]
  TOPIC Constructions =
    BASKET OID AS INTERLIS.UUIDOID;
    OID AS INTERLIS.STANDARDOID;
[...]
```

### In the data transferfile (`xtf`-file)...

... the OID of the *basket* is named `BID`

... the OID of the *objects* is named `TID`

```xml
<City_V1.Constructions BID="7dc3c035-b281-412f-9ba3-c69481054974">
  <City_V1.Constructions.Buildings TID="chMBAKER00000100"><Street>Rue des Fleures</Street><Number>1</Number></City_V1.Constructions.Buildings>
  <City_V1.Constructions.Buildings TID="chMBAKER00000101"><Street>Rue des Fleures</Street><Number>2</Number></City_V1.Constructions.Buildings>
```

### In the physical schema...

... those OIDs are written to the column `t_ili_tid`.

... while the `t_id` column contains just a ***schema-internal*** sequence used for foreign keys etc. Those `t_id`s are ***not*** the cross-system unique identificator.

![oid physical data](../assets/oid_physical_data.png)

!!! Note
    There are situations when that you can find the `t_id` as `TID` in your data. It's when there is no `t_ili_tid` available ili2db could use. But be aware those `TID`s are not stable then.

### OID Domains

> There are different types of OID domains. In Model Baker we try to set default values that fits into the definition.

Reviewer: ```suggestion
There are different types of OID domains. In Model Baker we try to set default values that fit into the definition.
```

- UUIDOID
- I32OID
- STANDARDOID
- ANYOID
- User defined OID

#### `UUIDOID`

It's defined as `OID TEXT*36` and needs to be an [Universally Unique Identifier (UUID)](https://datatracker.ietf.org/doc/html/rfc4122). While the probability that a UUID will be duplicated is not zero, it is generally considered close enough to zero to be negligible.

In the QGIS Project the default value expression for `t_ili_tid` is therefore this:
```
uuid('WithoutBraces')
```

#### `I32OID`

It's defined as `OID 0 .. 2147483647` what means it needs to be a positive 4 byte integer value.

As a counter we take the one provided by the `t_id` sequence.

In the QGIS Project the default value expression for `t_ili_tid` is therefore this:

```
t_id
```

#### `STANDARDOID`

It's defined as `OID TEXT*16` and follows some specific requirements.

> It requireds an 8 char prefix and 8 char postfix:

Reviewer: ```suggestion
It requires an 8 char prefix and 8 char postfix:
```

> - **Prefix (2 + 6 chars):** Country identifier + a *global identification part* assigned once by the [official authority](https://www.interlis.ch/dienste/oid-bestellen)

Reviewer: ```suggestion
- **Prefix (2 + 6 chars):** Country identifier + a *global identification part*. The global identification part can be ordered from the [official authority](https://www.interlis.ch/dienste/oid-bestellen)
```

- **Postfix (8 chars):** Sequence (numeric or alphanumeric) of your system as *local identification part*

> Model Baker does not know what your *global identification part* is, why it's important that you edit this part of the expression.

Author: ```suggestion
Model Baker does not know what your *global identification part* is and uses a test-prefix `ch100000`. It's important that replace this part with your own prefix. 
```

Author: @m-kuhn hope it's clearer now.

Reviewer: @signedav  test-prefix -> sample-prefix?

Reviewer 2: How about "registering" a real prefix with interlis.ch as a MB sample prefix? I might have the possibility to tweak the randomness of the prefix value... 😉

Reviewer: can we have `changeme`? :-)

Author: fine

Author: Oh I misunderstood.

`changeme` would be cool because it's descriptive, but only if we really can register it - otherwise it could lead to conflicts with the real owners of this prefix.

To register a Model Baker prefix is very undescriptive. It does not imply that you need to change it. And suddenly we have huge conflicts because people use the Model Baker prefix in their data.

Same with `ch100000` but at least this is defined as the test-prefix...

Author: See 
![image](https://github.com/opengisch/QgisModelBaker/assets/28384354/2d6a2628-f2d4-41f1-9910-26843a2d502b)

Reviewer 2: 😆 
`chANGEME` would at least emphasize the swissness of the default value...

seriously (and somehow out of scope of this PR, but anyway): a default/sample prefix will rarely be exchanged by the end user, because it will generate a completely valid transfer file. But if validating an exported transfer file raises OID errors because of a missing prefix, the user needs to address it (by inserting a prefix or even order one first at interlis.ch). So the awareness to get prefixes is raised when not having a valid fallback.

This contradicts my first comment, I know.

Author: Yes that is true. But I don't know if the goal should be that a new user creating the database and project and collects data already runs into validation issues after using the "suggested" default values.

Reviewer: Ideally they should be made aware before adding objects with OIDs.

Author: Not sure. Ideally yes, but often people just check out INTERLIS with the Model Baker as a first step / a learning and then I think they should not be confronted with STANDARDOID registration as a first step.

Then maybe it's better to create an invalid OID as preset value. Then they encounter it at the validation / export (and are still able to export the data invalid)...

Reviewer 2: Oh, I forgot about the recommendation in the reference manual!

I just got triggered by your sample/test prefix discussion and I am absolutely not into technical details of this PR. But in other commercial systems I know, the prefix is stored in a metadata table. When one changes it, it is updated for all data objects in the project. So the change can also be made after having objects already inserted into the db.

How about using a sample prefix that does not validate when used in an OID? Like `%bmpref%` or `%change%`?

Author: > But in other commercial systems I know, the prefix is stored in a metadata table. When one changes it, it is updated for all data objects in the project. So the change can also be made after having objects already inserted into the db.

Do you have more information about that?

> How about using a sample prefix that does not validate when used in an OID? Like %bmpref% or %change%?

Yes that's what I suggested. `_` or `%` as first character breaks the validity. `%change%` would be fine for me.

Reviewer 2: > Yes that's what I suggested. `_` or `%` as first character breaks the validity.

Sorry, I didn't see your comment about this because my browser didn't refresh the page. It's a good sign to have the same idea though :wink:

> `%change%` would be fine for me.

For me too! 👌 

> Do you have more information about that?

Well, this system isn't configured using ili2db so there are no constraints to stick to the provided INTERLIS schema.  But basically a table which contains objects uses the following column setup:

`prefix`, `sequence (project-wide)`, `OID (=concat(prefix+lpad(sequence)))`, `attribute1`, `attribute2` and so on.

The prefix source is located in a metadata table. Using SQL update, the prefix can be changed for the records and therefore the OID will be updated as well.

Author: Ah okay. That's a nice way to have it implemented in the back end. Not scope of this implementation, but good to know.

Concerning default value I'll go for %change% - in QGIS it's easy to replace all the %change% of a layer with a correct prefix.

If it will be often the case we might consider as well a follow up in the TID (OID) manager: "Replace prefix of STANDARDOID in all table with ..."

> Model Baker does not know what your *global identification part* is, why it's important that you edit this part of the expression.

Author: ```suggestion
Model Baker does not know what your *global identification part* is and uses a sample-prefix `ch100000`. It's important that replace this part with your own prefix.
```

> As *local identification part* Model Baker suggests the couter provided by the `t_id` sequence.

Author: ```suggestion
As *local identification part* Model Baker suggests the counter provided by the `t_id` sequence.
```

In the QGIS Project the default value expression for `t_ili_tid` is therefore this:

```
'ch100000' || lpad( T_Id, 8, 0 )
```

> 

Reviewer: ```suggestion

assuming `ch` as the country code and `100000` as the registered OID prefix.
```

Reviewer: or is `ch100000` the OID prefix and `100000` the "global identification part"?

#### `ANYOID`

> The `ANYOID` does not define a format of the OID but just that an OID needs to be defined in all the extended models. This domain is only used on topics that needs to be extended. There this sollution has some [limitations](#limitations).

Reviewer: ```suggestion
The `ANYOID` does not define a format of the OID but just that an OID needs to be defined in all the extended models. This domain is only used on topics that need to be extended. There this sollution has some [limitations](#limitations).
```

In the QGIS Project the default value expression for `t_ili_tid` is therefore this:

```
'_' || uuid('WithoutBraces')
```

#### Summary
| OID domain | Suggested default value expression |
|---|---|
| INTERLIS.UUIDOID |  `uuid('WithoutBraces')`  |
| INTERLIS.I32OID | `t_id` |
| INTERLIS.STANDARDOID |  `'ch100000' \|\| lpad( t_id, 8, 0 )` |
| INTERLIS.ANYOID | `'_' \|\| uuid('WithoutBraces')` |
| City_V1.TypeID | `'_' \|\| uuid('WithoutBraces')` |
| not defined | `'_' \|\| uuid('WithoutBraces')` |

### TID (OID) Manager

Since the user have to be able to edit those values, they are provided in the GUI.

Additionally to the [page implemented in the wizard](../../user_guide/import_workflow/#tid_(oid)_values) the configuration of the TIDs can be made on existing QGIS Projects.

Find the **TID (OID) Manager** via the *Database > Model Baker* menu.

![tid manager](../assets/oid_tid_manager.png)

Here you can use the QGIS Expression Dialog to edit the default value expression for the `t_ili_tid` field of each layer.

If you want to have the `t_ili_tid` exposed to the form, you can select ***Show***.

> If you need a counter in the expressions, you can use the `t_id` field, that has a schema-wide sequence counting up. This sequence can be reset as well by the user, but be carefull not to set it lower than already existing `t_id`s in your project. See below the [limitations](#limitations).

Reviewer: ```suggestion
If you need a counter in the expressions, you can use the `t_id` field, that has a schema-wide sequence counting up. This sequence can be reset as well by the user, but be careful not to set it lower than already existing `t_id`s in your project. See below the [limitations](#limitations).
```

### Limitations

> This sollution covers a lot, but not everything.

Reviewer: ```suggestion
This solution covers a lot, but not everything.
```
